librz/bin/bobj.c

```diff
@@ -1017,3 +995,119 @@ RZ_API ut64 rz_bin_object_v2p(RzBinObject *obj, ut64 vaddr) {
 	}
 	return m->paddr + delta;
 }
+
+RZ_IPI RZ_OWN RzBinStrDb *rz_bin_string_database_new(RzList *list, ut64 load_address, bool decode_base64) {
+	RzBinStrDb *db = RZ_NEW0(RzBinStrDb);
+	if (!db) {
+		RZ_LOG_ERROR("rz_bin: Cannot allocate RzBinStrDb\n");
+		rz_list_free(list);
+		return NULL;
+	}
+
+	db->list = list ? list : rz_list_newf((RzListFree)rz_bin_string_free);
+	db->phys = ht_up_new0();
+	db->virt = ht_up_new0();
+	if (!db->list || !db->phys || !db->virt) {
+		RZ_LOG_ERROR("rz_bin: Cannot allocate RzBinStrDb internal data structure.\n");
+		goto fail;
+	}
+
+	RzListIter *it;
+	RzBinString *bstr;
+	rz_list_foreach (list, it, bstr) {
+		bstr->paddr += load_address;
+		if (decode_base64) {
+			char *decoded = bstr->string;
+			do {
+				// ensure to decode base64 strings encoded multiple times.
+				char *tmp = (char *)rz_base64_decode_dyn(decoded, -1);
+				if (!tmp || !rz_str_is_printable(tmp)) {
+					free(tmp);
+					break;
+				}
+				free(decoded);
+				decoded = tmp;
+			} while (1);
+			if (decoded != bstr->string) {
+				free(bstr->string);
+				bstr->string = decoded;
+				bstr->type = RZ_STRING_ENC_BASE64;
+			}
+		}
+
+		if (!ht_up_update(db->phys, bstr->paddr, bstr)) {
+			RZ_LOG_ERROR("rz_bin: Cannot insert RzBinString in RzBinStrDb (phys)\n");
+			goto fail;
+		}
+		if (!ht_up_update(db->virt, bstr->vaddr, bstr)) {
+			RZ_LOG_ERROR("rz_bin: Cannot insert RzBinString in RzBinStrDb (virt)\n");
+			goto fail;
+		}
+	}
+	return db;
+
+fail:
+	rz_bin_string_database_free(db);
+	return NULL;
+}
+
+RZ_IPI void rz_bin_string_database_free(RzBinStrDb *db) {
+	if (!db) {
+		return;
+	}
+	rz_list_free(db->list);
+	ht_up_free(db->phys);
+	ht_up_free(db->virt);
+	free(db);
+}
+
+static int string_compare_sort(const RzBinString *a, const RzBinString *b) {
+	if (b->paddr > a->paddr) {
+		return -1;
+	} else if (b->paddr < a->paddr) {
+		return 1;
+	} else if (b->vaddr > a->vaddr) {
+		return -1;
+	} else if (b->vaddr < a->vaddr) {
+		return 1;
+	}
+	return 0;
+}
+
+RZ_IPI bool rz_bin_string_database_add(RzBinStrDb *db, RzBinString *bstr, ut64 load_address, bool decode_base64) {
+	rz_return_val_if_fail(db, false);
+
+	bstr->paddr += load_address;
+	if (decode_base64) {
+		char *decoded = bstr->string;
+		do {
+			// ensure to decode base64 strings encoded multiple times.
+			char *tmp = (char *)rz_base64_decode_dyn(decoded, -1);
+			if (!tmp || !rz_str_is_printable(tmp)) {
+				free(tmp);
+				break;
+			}
+			free(decoded);
+			decoded = tmp;
+		} while (1);
+		if (decoded != bstr->string) {
+			free(bstr->string);
+			bstr->string = decoded;
+			bstr->type = RZ_STRING_ENC_BASE64;
+		}
+	}
+
+	if (!rz_list_add_sorted(db->list, bstr, (RzListComparator)string_compare_sort)) {
```

Comment:
This is horribly slow due to linear search. Two possibilities for different data structures would be:
* Sorted Rz(P)Vector, super fast access, but slower insertion
* RBTree, a bit more indirection, but fast insertion
For reasons I am to lazy to write down right now, I am leaning towards the RBTree, but in this pr I would leave it as-is since this is no regression here. So just see this as a general comment.

Reply:
+1 for RBTree